### PR TITLE
Fix space leak on cradle reloads

### PR DESCRIPTION
**Pull request by pepeiborra**
We were leaking 'PackageExports' values for all past 'HscEnv's. 
Fixed by moving the package exports inside the 'HscEnvEq'.

I have also added two experiments to characterise the leak, the main one is 'code actions after cradle edit'. On the Cabal 3 example, this is what the experiment looks like:

image

### CODE REVIEW

File: ghcide/src/Development/IDE/Plugin/CodeAction.hs [hunk 1]
```diff
@@ -1090,10 +1090,10 @@ constructNewImportSuggestions
 constructNewImportSuggestions exportsMap (qual, thingMissing) notTheseModules = nubOrd
   [ suggestion
   | Just name <- [T.stripPrefix (maybe "" (<> ".") qual) $ notInScope thingMissing]
-  , (identInfo, m) <- maybe [] Set.toList $ Map.lookup name (getExportsMap exportsMap)
+  , identInfo <- maybe [] Set.toList $ Map.lookup name (getExportsMap exportsMap)
   , canUseIdent thingMissing identInfo
-  , m `notElem` fromMaybe [] notTheseModules
-  , suggestion <- renderNewImport identInfo m
+  , moduleNameText identInfo `notElem` fromMaybe [] notTheseModules
+  , suggestion <- renderNewImport identInfo (moduleNameText identInfo)
```
Comment (Ailrun) [inline ghcide/src/Development/IDE/Plugin/CodeAction.hs:1096]: Do we need to pass `(moduleNameText identInfo)` part? That's now redundant as it is in `identInfo`.

**Review by Ailrun — APPROVED** [review]
LGTM

File: ghcide/src/Development/IDE/Plugin/CodeAction.hs (hunk 1 above)
Comment (pepeiborra) [reply ghcide/src/Development/IDE/Plugin/CodeAction.hs]: well spotted, thanks

**Comment by wz1000** [comment]
Now that the `hiedb` changes have landed, I think this feature would be a good candidate to implement using the database.

**Comment by pepeiborra** [comment]
> Now that the `hiedb` changes have landed, I think this feature would be a good candidate to implement using the database.

But hiedb doesn't index build depends, does it?

**Comment by wz1000** [comment]
> But hiedb doesn't index build depends, does it?

Ah, right. I misread the PR and thought it was for the project local exports map.

However, we could still index and persist a export map for dependent packages, by indexing the exports directly from the ModIface instead of the `.hie` file.

**Comment by pepeiborra** [comment]
All the tests pass now, but the benchmark run on master (where the leak is still present) times out. I'll wait for the speed ups in #1315 to land

**Comment by pepeiborra** [comment]
> > But hiedb doesn't index build depends, does it?
> 
> Ah, right. I misread the PR and thought it was for the project local exports map.
> 
> However, we could still index and persist a export map for dependent packages, by indexing the exports directly from the ModIface instead of the `.hie` file.

That's how the exports map works now, except that it isn't persisted. Do you think that persisting would be worth the effort? I think that's a nice idea, please file an issue!

**Pull request merged**